### PR TITLE
Make it possible to run individual pytest cases with test.py

### PR DESCRIPTION
**Pull request by xemul**
Today's test.py allows filtering tests to run with the `test.py --options name` syntax. The "name" argument is then considered to be some prefix, and when iterating tests only those whose name starts with that prefix are collected and executed. This has two troubles.

Minor: since it is prefix filtering, running e.g. topology_custom/test_tablets will run test_tablets _and_ test_tablets_migration from it. There's no way to exclude the latter from this selection. It's not common, but careful file names selection is welcome for better ~~user~~ testing experience.

Major: most of test files in topology and python suites contain many cases, some are extremely long. When the intent is to run a single, potentially fast, test case one needs to either wait or patch the test .py file by hand to somehow exclude unwanted test cases.

This PR adds the ability to run individual test case with test.py. The new syntax is `test.py --options name::case`. If the "::case" part is present two changes apply.

First, the test file selection is done by name match, not by prefix match. So running topology_custom/test_tablets will _not_ select test_tablets_migration from it.

Second, the "::case" part is appended to the pytest execution so that it collects and runs only the specified test case.

### CODE REVIEW

File: test.py
Comment (nyh) [inline test.py:932]: A suggestion: for the benefit of pytest fanboys like myself, it would be nice to also support "name.py::case" and not just "name::case". I think it should be easy, just by not adding ".py" if there's already one?

**Comment by denesb** [comment]
> This PR adds the ability to run individual test case with test.py. The new syntax is `test.py --options name::case`. If the "::case" part is present two changes apply.
> 
> First, the test file selection is done by name match, not by prefix match. So running topology_custom/test_tablets will _not_ select test_tablets_migration from it.

I think this is already better than what we have. But I think it will be slightly confusing and somewhat incomplete. Why not go all the way and allow full match with just a filename? We could add `--prefix` and `--exact`, with the latter supporting test cases too. The remaining question is then what should be the default, when only positional filter is provided: prefix or exact.

That said, I'm fine with merging this PR as-is and addressing this in a follow-up.

**Comment by lkshminarayanan** [comment]
> /cc @lkshminarayanan who was working on something similar.

@denesb yes, I was doing something similar - support for formats like `suite::name::case`,  `name::case`, `suite::` in the `name` option. But this looks better and I can continue on top of this if required.

**Comment by xemul** [comment]
> What is the "--options" thing in your examples in the commit message? 

It's any option test.py accepts. The `...` as you proposed below would be simpler indeed.

> I always ran test.py name and it worked, there is no need for the --options flag. Or... maybe... you just wrote --options to signify that you can add more options? :-) In this case, just an elipses (...) is absolutely nothing would have been clearer.
> 
> If I understand correctly, the implementation is incomplete because it's only supported in pytest-based tests. 

Yes. There are also boost tests that accept `--run-test` option, but it's already in use with `--parallel-cases` one and needs more care. I planned to do it as a followup since boost test can be mostly run by hand with `./build/<mode>/test/boost/<test> --run-test=<test>` syntax, while topology test cannot.

> Finally, please document the new syntax in docs/dev/testing.md which is the documentation for test.py.

Sure!

> I don't insist on any of this as a prerequisite to merging this PR - I think it can be improved incrementally later.

Will do

**Review by denesb — APPROVED** [review]

**Comment by avikivity** [comment]
How do I know what name::case matches my test?

**Comment by xemul** [comment]
> How do I know what name::case matches my test?

For pytests the name is the filename (e.g. test_tablets.py) and case is a function in that file starting with `test_` e.g. test_tablet_replication_factor_enough_nodes. For boost tests name is, again, the filename and case name is top-level function wrapped with BOOST_AUTO_TEST_CASE or ,SEASTAR_TEST_CASE, they are listed with `--list_content` option

**Comment by avikivity** [comment]
I think we should report to stdout the same name that we match. Otherwise it's a guessing game.

**Comment by xemul** [comment]
> I think we should report to stdout the same name that we match. Otherwise it's a guessing game.

It already works like that. E.g.

```
topology_custom release [ PASS ] topology_custom.test_tablets_migration.1 0.27s
^                                ^               ^
suite                            name            case
```

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by xemul** [comment]
@nyh , please consider merging

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

File: test.py
Comment (kostja) [inline test.py:282]: I think this change makes the code very hairy. You don't need to split the pattern if it's not set. You also don't recognize skip_pattern option. What you could be doing instead is an additional check of options.patterns *inside* PythonTestSuite::add_test, so that if the pattern is *provided (there is no point in splititng and checking if there is no pattern), then it is additionally studied for the case name. 

This would spare you from having to pass case name through entire add_test hierarchy.

**Comment by kostja** [comment]
Conclusion: as long as any server maintainer can merge a patch to test.py, test.py code will continue its rot :/

**Comment by nyh** [comment]
> Conclusion: as long as any server maintainer can merge a patch to test.py, test.py code will continue its rot :/

Did you post a negative review ("request changes")? That's what you should do if you want to officially express that you're against this patch.

**Comment by nyh** [comment]
> > Conclusion: as long as any server maintainer can merge a patch to test.py, test.py code will continue its rot :/
> 
> Did you post a negative review ("request changes")? That's what you should do if you want to officially express that you're against this patch.

To clarify what I meant: Two maintainers reviewed this PR positively, another developer did this too, and a third maintainer merged it. So it's not "one server maintainer", it's actually four people. You yourself didn't mark "request changes", just a "comment" so it wasn't entirely clear that you were against the patch in its current form. Had you said "request changes", I'm pretty sure the third maintainer who merged this PR wouldn't have done so until you removed your nack - even though four other people acked. But you didn't.

Finally, everything you said is obviously true not just for test.py, but also for db/view/view.cc or alternator/executor.cc. So perhaps a correct point that you're raising isn't really about the number of reviewers needed to approve a patch, but about their identity: Maybe it makes sense that if one developer is strongly invested in a piece of code - like you are in test.py, or me in alternator/executor.cc - we should give that developer the courtesy of reviewing changes to this code, and not merge until an explicit "approve" from this person - even if four other people already approved. We often do this (ask a specific person to review before merging), but not always. I don't think this can ever be a rule (we can't delay merging for weeks if the "owner" of the code is slow to review it and several other people reviewed), but I guess it can be something we make a better effort to do.

**Comment by denesb** [comment]
> Conclusion: as long as any server maintainer can merge a patch to test.py, test.py code will continue its rot :/

Also, you posted your review after the PR was queued -- it just took a lot of time for the promotion to happen.

**Comment by kostja** [comment]
@nyh @xemul @denesb - I am asking @xtrey to work on the follow -ups

**Pull request merged**
